### PR TITLE
Made commander do averaging and run a variable amount of iterations

### PR DESCRIPTION
**Pull request by JGelfand**
~~TODO: make commander not use a global results variable, and/or make the results shared somehow~~

I'm making the pr now so you don't end up doing the same work. I'll finish up with the global stuff tomorrow.

### CODE REVIEW

**Comment by JGelfand** [comment]
Should be good now

File: scripts/commander.py [hunk 1]
```diff
@@ -216,37 +227,37 @@ def main_helper(N=1000, partitions=[], path_prefix='./workspace/parric-ttmm/ttmm
...
-    main_rec(0, parent, partitions, path_prefix, keep)
+    main_rec(0, parent, partitions, path_prefix, keep, iterations)
 
 
-def main_rec(level, parent, partitions, path_prefix, keep):
+def main_rec(level, parent, partitions, path_prefix, keep, iterations):
     global master_cube_N
     
-    if level == 3:
+    if level >= iterations:
```
Comment (lnarmour) [inline scripts/commander.py]: Haha the real basecase, there it is.

File: scripts/commander.py [hunk 2]
```diff
@@ -216,37 +227,37 @@ def main_helper(N=1000, partitions=[], path_prefix='./workspace/parric-ttmm/ttmm
...
+    results={}
     # Add tasks to queue
     tasks = queue.Queue()
     print('\nCreating tasks for parent', parent.c,'...')
     for child in partition(parent, partitions[level]):
         (ts1, ts2, ts3) = child.center()
-        tasks.put(Command('{}/TMM'.format(path_prefix), [master_cube_N, ts1, ts2, ts3]))
-    run_workers(machines, tasks, level, parent)
+        for i in range(5):
+            tasks.put(Command('{}/TMM'.format(path_prefix), [master_cube_N, ts1, ts2, ts3]))
```
Comment (lnarmour) [inline scripts/commander.py]: Okay, so we just execute each command 5 times now.

File: scripts/commander.py [hunk 3]
```diff
@@ -216,37 +227,37 @@ def main_helper(N=1000, partitions=[], path_prefix='./workspace/parric-ttmm/ttmm
...
+    run_workers(machines, tasks, level, results, parent)
     print('...done.')
-
-    # choose best results for next iteration, i.e. next set of parents
-    results[level].sort()
-    parents = []
     
+    # choose best results for next iteration, i.e. next set of parents
+    sorted_results=[(results[command], command) for command in results]
```
Comment (lnarmour) [inline scripts/commander.py]: Okay, so now sorted_results is a 2-tuple of a "list of the 5 to-be-averaged Result objects" and the "Command object".

File: scripts/commander.py [hunk 4]
```diff
@@ -216,37 +227,37 @@ def main_helper(N=1000, partitions=[], path_prefix='./workspace/parric-ttmm/ttmm
...
+    sorted_results=sorted([ ( statistics.mean([r.time for r in sr[0]]), sr[1]) for sr in sorted_results])
```
Comment (lnarmour) [inline scripts/commander.py]: And then here you calc the mean time of the "list of the 5 Result objects".  Then sort them.

**Comment by lnarmour** [comment]
@JGelfand 
Was testing with different inputs and this one throws a TypeError:
```
c.main_helper(N=100, partitions=[4,3,2], iterations=3)
```
Here's the stack trace, I'm looking at it now but maybe you'll notice something quicker:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/s/bach/g/under/lnarmour/workspace/joseph/parric-ttmm/scripts/commander.py", line 230, in main_helper
    main_rec(0, parent, partitions, path_prefix, keep, iterations)
  File "/s/bach/g/under/lnarmour/workspace/joseph/parric-ttmm/scripts/commander.py", line 266, in main_rec
    main_rec(level+1, Cube(tuple(r[1].params[1:]), parent.l//partitions[level]), partitions, path_prefix, keep, iterations)
  File "/s/bach/g/under/lnarmour/workspace/joseph/parric-ttmm/scripts/commander.py", line 255, in main_rec
    sorted_results=sorted([ ( statistics.mean([r.time for r in sr[0]]), sr[1]) for sr in sorted_results])
  File "/s/bach/g/under/lnarmour/workspace/joseph/parric-ttmm/scripts/commander.py", line 137, in __lt__
    return self.__hash__ < other.__hash__
TypeError: '<' not supported between instances of 'method' and 'method'
```

**Comment by lnarmour** [comment]
Yup, that TypeError from previous comment doesn't occur if you change N to 1000.
So this does NOT throw that error:
```
c.main_helper(N=1000, partitions=[4,3,2], iterations=3)
```

**Comment by JGelfand** [comment]
Sorry, just noticed that you were responding. The type error is due to tuple comparison using the next layer of tuple for comparisons if the initial is the same. I tried to fix it by adding an __lt__ function to Command, but I made it compare __hash__ instead of __hash__(). I'm going to fix it and upload a separate PR.

**Pull request merged**